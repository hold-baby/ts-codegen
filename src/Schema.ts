import { getUseExtends, isArray, toCapitalCase } from "src/utils";
import { forEach, isEmpty, map, reduce } from "lodash";
import { IReference, ISchema } from "src/__types__/OpenAPI";
import { CustomType, Type } from "src/Type";
import { CustomSchema } from "src/__types__/types";
import { createRegister } from "src/Register";

export class Schema {
  private type: Type;

  constructor(register: ReturnType<typeof createRegister>) {
    this.type = new Type(register);
  }

  public convert(schema: CustomSchema, id?: string): CustomType {
    const name = id ? toCapitalCase(id) : id;
    const oneOf = (schema as ISchema).oneOf || (schema as ISchema).anyOf;
    if (oneOf) {
      return this.type.oneOf(map(oneOf, (v) => this.convert(v)));
    }

    const allOf = (schema as ISchema).allOf;
    if (allOf) {
      const { props, refs, useExtends } = this.handleAllOf(allOf, name);
      return this.type.object(props, refs, useExtends);
    }

    if (schema.items) {
      return this.type.array(this.handleItems(schema.items, name));
    }

    if (schema.$ref) {
      return this.type.ref((schema as IReference).$ref);
    }

    if (schema.enum) {
      return this.type.enum(schema.enum, name);
    }

    if (schema.type === "object" || schema.properties) {
      return schema.properties ? this.type.object(this.handleObject(schema, name)) : this.type.object("object");
    }

    if (schema.type === "string") {
      return this.type.string();
    }

    if (schema.type === "boolean") {
      return this.type.boolean();
    }

    if (schema.type === "integer" || schema.type === "number") {
      return this.type.number();
    }

    if (schema.type === "file") {
      return this.type.file();
    }

    return this.type.null();
  }

  private handleAllOf(schemas: Array<CustomSchema>, name?: string) {
    const refs: any[] = [];
    let props: any = {};
    let useExtends = getUseExtends(schemas);

    forEach(schemas, (schema) => {
      if (schema.$ref) {
        refs.push(this.convert(schema, name));
      } else if (!isEmpty(schema)) {
        props = this.convert(schema, name);
      }
    });

    return {
      refs,
      props,
      useExtends,
    };
  }

  private handleItems(items: CustomSchema | IReference | CustomSchema[], name?: string): CustomType | CustomType[] {
    if (isArray(items)) {
      return map(items, (v) => this.handleItems(v, name)) as CustomType[];
    }
    return this.convert(items, name);
  }

  private handleObject(schema: CustomSchema, name?: string): { [key: string]: CustomType } {
    return reduce(
      schema.properties,
      (res, v, k) => {
        const isRequired = (v as CustomSchema)?.required || schema.required?.includes(k);
        return {
          ...res,
          [`${k}${isRequired ? "" : "?"}`]: this.convert(v, `${name}${toCapitalCase(k)}`),
        };
      },
      {},
    );
  }
}
