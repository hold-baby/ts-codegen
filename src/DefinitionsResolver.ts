import { Schema } from "swagger-schema-official";
import { addPrefixForInterface, arrayToObject, isNumber, toCapitalCase, toTypes } from "./utils";
import { compact, Dictionary, forEach, includes, replace, some } from "lodash";
import { SchemaResolver2 } from "src/swagger/v3/SchemaResolver2";

// TODO: 1. Handle required params.
// TODO: handle `in: fromData`
// TODO: handle `in schema`
// TODO: 确认不同 endpoint 是否都会生成 swagger

const ENUM_SUFFIX = `#EnumTypeSuffix`;

export function generateEnums(data: Dictionary<any>, key: string) {
  if (!data) {
    return "";
  }

  const enums = data[key];
  const hasNumber = some(enums, (v) => isNumber(v));
  const enumName = replace(key, ENUM_SUFFIX, "");
  return hasNumber
    ? `export type ${enumName} = ${enums.map((item: string | number) => JSON.stringify(item)).join("|")}`
    : `export enum ${enumName} ${JSON.stringify(arrayToObject(enums)).replace(/:/gi, "=")}`;
}

export class DefinitionsResolver {
  public resolvedDefinitions: Dictionary<any> = {};

  static of(definitions?: { [definitionsName: string]: Schema }) {
    return new DefinitionsResolver(definitions);
  }

  constructor(private definitions?: { [definitionsName: string]: Schema }) {}

  scan = () => {
    const r = SchemaResolver2.of((k, v) => {
      this.resolvedDefinitions[k] = v;
    });

    forEach(this.definitions, (v, k) =>
      r.resolve({
        ...v,
        _name: k,
        _propKey: k,
      }),
    );

    return this;
  };

  toDeclarations = (): string[] => {
    const resolvedDefinitions = this.resolvedDefinitions;
    const arr = Object.keys(resolvedDefinitions)
      .sort()
      .map((key) => {
        if (includes(key, ENUM_SUFFIX)) {
          return generateEnums(resolvedDefinitions, key);
        }

        if (resolvedDefinitions[key] === "object") {
          return `export interface ${addPrefixForInterface(toCapitalCase(key))} {[key:string]:any}`;
        }
        const val = toTypes(resolvedDefinitions[key]);
        if (val) {
          return `export interface ${addPrefixForInterface(toCapitalCase(key))} ${val}`;
        }
      });

    return compact(arr);
  };
}
