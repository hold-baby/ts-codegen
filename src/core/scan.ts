import { IReference } from "src/v3/OpenAPI";
import { CustomSchema, Ref, Register } from "src/core/Type";
import { keys } from "lodash";
import { getUseExtends, Schema } from "src/core/Schema";
import { prettifyCode, toCapitalCase } from "src/core/utils";

const getDeclarationType = (schema: CustomSchema) => {
  if (schema.type === "object" || schema.properties || (schema.allOf && getUseExtends(schema.allOf))) {
    return "interface";
  }
  return "type";
};

const addPrefix = (name: string) => `${Register.prefixes[name] === "interface" ? "I" : "T"}${name}`;

export const scan = (schemas: { [k: string]: CustomSchema | IReference }) => {
  const schemaHandler = new Schema();

  keys(schemas).forEach((k) => {
    const name = toCapitalCase(k);
    const type = schemaHandler.convert(schemas[k], name);
    Register.setType(name, type);
    Register.setPrefix(name, getDeclarationType(schemas[k]));
  });

  for (let name in Register.refs) {
    (Register.refs[name] as Ref).rename(addPrefix(name));
  }

  let output = "";
  keys(Register.decls)
    .sort()
    .forEach((k) => {
      output =
        output +
        `export ${Register.prefixes[k]} ${addPrefix(k)} ${
          Register.prefixes[k] === "interface" ? "" : "="
        } ${Register.decls[k].toType()}${Register.prefixes[k] === "type" ? ";" : ""}\n\n`;
    });

  return prettifyCode(output);
};
