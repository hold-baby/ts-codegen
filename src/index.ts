export * from "src/v2/ClientConfigs";
export * from "src/v3/ClientConfigsV3";
export * from "src/core/utils";
