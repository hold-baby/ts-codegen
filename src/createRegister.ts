// 利用闭包持有状态（私有变量）
import { CustomType, Ref } from "src/Type";
import { Parameter } from "swagger-schema-official";
import { IReference, IRequestBody, IResponse } from "src/__types__/OpenAPI";
import { get, set } from "lodash";

export enum DeclKinds {
  interface = "interface",
  type = "type",
  enum = "enum",
}

const withPrefix = (name: string, kind: string) => {
  switch (kind) {
    case "interface":
      return `I${name}`;
    case "type":
      return `T${name}`;
    default:
      return name;
  }
};

export interface IStore {
  decls: {
    [id: string]: {
      type: CustomType;
      kind: string;
      name: string;
    };
  };
  refs: { [id: string]: Ref };
  parameters: {
    [id: string]: Parameter;
  };
  responses: {
    [id: string]: Response | IResponse | IReference;
  };
  requestBodies: {
    [id: string]: IReference | IRequestBody;
  };
}

export const createRegister = () => {
  const store: IStore = {
    decls: {},
    refs: {},
    parameters: {},
    responses: {},
    requestBodies: {},
  };

  return {
    getDecls() {
      return store.decls;
    },

    setDecl: (id: string, type: CustomType, kind: string) => {
      store.decls[id] = {
        type,
        kind,
        name: withPrefix(id, kind),
      };
    },

    setRef: (id: string) => {
      if (store.refs[id]) {
        return store.refs[id];
      }

      const type = new Ref(id);
      store.refs[id] = type;

      return type;
    },

    getData(paths: string[]) {
      return get(store, paths);
    },

    setData(paths: string[], data: any) {
      return set(store, paths, data);
    },

    renameAllRefs: (cb: (newName: string) => string) => {
      for (let name in store.refs) {
        store.refs[name].rename(cb(name));
      }
    },
  };
};
