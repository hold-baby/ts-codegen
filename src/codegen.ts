import { getFilename, prettifyCode, testJSON } from "src/utils";
import axios from "axios";
import { isEmpty, map } from "lodash";
import * as fs from "fs";
import * as path from "path";
import { IOpenAPI } from "src/__types__/OpenAPI";
import { Spec } from "swagger-schema-official";
import { getInputs, scan } from "src/scan";
import { ERROR_MESSAGES } from "src/constants";

export const getCodegenConfig = () => {
  const codegenConfigPath = path.resolve("ts-codegen.config.json");
  return fs.existsSync(codegenConfigPath)
    ? require(codegenConfigPath)
    : {
        output: ".output",
        actionCreatorImport: "",
        clients: [],
        options: {
          typeWithPrefix: false,
          backwardCompatible: false,
        },
      };
};

export const codegen = () => {
  const { output, actionCreatorImport, timeout, data, clients, options } = getCodegenConfig();

  const writeSpecToFile = (spec: IOpenAPI | Spec) => {
    if (!spec) {
      return;
    }
    const fileStr = `${actionCreatorImport} ${scan(spec, options)}`;
    const { basePath } = getInputs(spec);
    write(output, `./${getFilename(basePath)}`, fileStr);
  };

  if (!isEmpty(data)) {
    data.map((file: string) => {
      const specStr = fs.readFileSync(file, "utf8");
      const spec = testJSON(specStr, ERROR_MESSAGES.INVALID_JSON_FILE_ERROR);

      writeSpecToFile(spec);
    });
  }

  if (!isEmpty(clients)) {
    fetchSwaggerJSON(clients, timeout).then((results: any[]) => {
      results.forEach((spec: IOpenAPI | Spec) => {
        writeSpecToFile(spec);
      });
    });
  }
};

const write = (output: string, filename: string, str: string) => {
  if (!fs.existsSync(output)) {
    fs.mkdirSync(output);
  }

  fs.writeFileSync(path.resolve(output, `./${filename}.ts`), prettifyCode(str), "utf-8");
};

const fetchSwaggerJSON = (clients: string[] = [], timeout: number = 10 * 1000) => {
  const instance = axios.create({
    timeout,
  });

  return Promise.all(
    map(clients, (client) => {
      instance
        .get(client)
        .then((response) => response.data)
        .catch((error) => {
          console.error(`${error.code}: ${ERROR_MESSAGES.FETCH_CLIENT_FAILED_ERROR}`);
        });
    }),
  );
};
