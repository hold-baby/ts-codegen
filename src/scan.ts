import { IOpenAPI, IServer } from "src/__types__/OpenAPI";
import { CustomType } from "src/Type";
import { compact, get, isEmpty, keys, mapValues, sortBy } from "lodash";
import { Schema } from "src/Schema";
import { getUseExtends, prettifyCode, setDeprecated, toCapitalCase, toTypes } from "src/utils";
import { Spec } from "swagger-schema-official";
import { CustomReference, CustomSchema, IClientConfig, RequestType } from "src/__types__/types";
import { createRegister, DeclKinds, IStore } from "src/createRegister";
import { parse } from "url";
import { getClientConfigsV2, getClientConfigsV3 } from "src/createClientConfigs";

enum DataType {
  openapi,
  swagger,
}

interface ScanOptions {
  typeWithPrefix?: boolean; // Will keep prefix('I' for interface, 'T' for type) in types when it sets true
  backwardCompatible?: boolean; // Not recommend, only if you want backward capability. This option will help to keep operationId and method name as before when it sets true. This option is only worked with swagger version 2.0.
}

export const scan = (data: Spec | IOpenAPI, options?: ScanOptions) => {
  const register = createRegister(options?.typeWithPrefix);
  const schemaHandler = new Schema(register);
  const { dataType, basePath, paths, schemas, parameters, responses, requestBodies } = getInputs(data);

  keys(schemas).forEach((k) => {
    const name = toCapitalCase(k);
    register.setDecl(name, schemaHandler.convert(schemas[k], name), getDeclarationType(schemas[k]));
  });

  register.setData(["parameters"], parameters);
  register.setData(["responses"], responses);
  register.setData(["requestBodies"], requestBodies);

  let clientConfigs: IClientConfig[] =
    dataType === DataType.swagger
      ? getClientConfigsV2(paths, basePath, register, options?.backwardCompatible)
      : getClientConfigsV3(paths, basePath, register);

  const decls = register.getDecls();
  if (options?.typeWithPrefix) {
    register.renameAllRefs((key) => decls[key].name);
  }

  return print(clientConfigs, decls);
};

const isOpenApi = (v: any): v is IOpenAPI => v.openapi;

export const getInputs = (data: Spec | IOpenAPI) => {
  if (isOpenApi(data)) {
    return {
      dataType: DataType.openapi,
      basePath: getBasePathFromServers(data?.servers),
      paths: data.paths,
      schemas: data.components?.schemas as { [key: string]: CustomSchema | CustomReference },
      parameters: data.components?.parameters,
      responses: data.components?.responses,
      requestBodies: data.components?.requestBodies,
    };
  }
  return {
    dataType: DataType.swagger,
    basePath: data.basePath || "",
    paths: data.paths,
    schemas: data.definitions as { [key: string]: CustomSchema },
    parameters: data.parameters,
    responses: data.responses,
    requestBodies: null,
  };
};

const print = (clientConfigs: IClientConfig[], decls: IStore["decls"]) => {
  return prettifyCode(`${printRequest(clientConfigs)} \n\n ${printTypes(decls)}`);
};

function printRequest(clientConfigs: IClientConfig[]): string {
  const configs = sortBy(clientConfigs, (o) => o.operationId);

  return configs
    .map((v) => {
      const toUrl = () => `url: \`${v.url}\`,`;
      const toMethod = () => `method: "${v.method}",`;
      const toRequestBody = () => {
        if (!isEmpty(v.bodyParams)) {
          // TODO: refactor code
          return `data: ${v.bodyParams!.length > 1 ? `{${v.bodyParams!.join(",")}}` : v.bodyParams![0]},`;
        }
        return v.contentType ? "data: requestBody," : "";
      };
      const toQueryParams = () => {
        const params = toRequestParams(v.queryParams);
        return params ? `params: ${params},` : "";
      };
      const toHeaders = () => (v.contentType ? `headers: {"Content-Type": '${v.contentType}'},` : "");
      const toGenerators = () => {
        const types = compact([generateTReq(v.TReq), v.TResp?.toType(false)]).join(",");
        return types ? `<${types}>` : "";
      };
      const toRequestInputs = () => {
        const getRequestBody = () => {
          if (!isEmpty(v.bodyParams)) {
            return v.bodyParams!;
          }
          return v.contentType ? ["requestBody"] : "";
        };
        const list = compact([...v.pathParams, ...v.queryParams, ...getRequestBody()]);
        return isEmpty(list) ? "" : toRequestParams(list);
      };

      return `
${v.deprecated ? setDeprecated(v.operationId) : ""}
export const ${v.operationId} = createRequestAction${toGenerators()}("${
        v.operationId
      }", (${toRequestInputs()}) => ({${toUrl()}${toMethod()}${toRequestBody()}${toQueryParams()}${toHeaders()}})
);
`;
    })
    .join("\n\n");
}

const printTypes = (decls: IStore["decls"]): string => {
  return keys(decls)
    .sort()
    .map((k) => {
      const expr = decls[k].kind === DeclKinds.type ? "=" : "";
      const semi = decls[k].kind === DeclKinds.type ? ";" : "";
      return `export ${decls[k].kind} ${decls[k].name} ${expr} ${decls[k].type.toType()}${semi}`;
    })
    .join("\n\n");
};

function generateTReq(TReq: IClientConfig["TReq"]) {
  if (isEmpty(TReq)) {
    return "";
  }

  function gen(obj: IClientConfig["TReq"]): string {
    return toTypes(
      mapValues(obj, (v) => {
        if (!v.toType) {
          return gen(v as RequestType);
        }
        return (v as CustomType).toType(false);
      }),
    );
  }

  return gen(TReq);
}

const toRequestParams = (data: string[]) =>
  !isEmpty(data)
    ? `{
${data.join(",\n")}
}`
    : undefined;

const getDeclarationType = (schema: CustomSchema) => {
  if (schema?.type === "object" || schema?.properties || (schema?.allOf && getUseExtends(schema?.allOf))) {
    return DeclKinds.interface;
  }
  return DeclKinds.type;
};

const getBasePathFromServers = (servers?: IServer[]): string => {
  if (isEmpty(servers)) {
    return "";
  }
  const server = servers![0];
  if (server?.variables) {
    const basePath = get(server, "variables.basePath.default");
    return basePath ? `/${basePath}` : "";
  }
  return parse(server?.url)?.pathname || "";
};
