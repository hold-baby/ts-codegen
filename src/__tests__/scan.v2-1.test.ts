import swaggerV2 from "examples/swagger.v2.json";
import { Spec } from "swagger-schema-official";
import { scan } from "src/scan";

describe("swagger v2", () => {
  it("should handle spec correctly", () => {
    expect(scan(swaggerV2 as Spec)).toMatchSnapshot();
  });
});
