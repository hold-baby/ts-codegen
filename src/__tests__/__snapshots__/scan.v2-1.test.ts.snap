// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swagger v2 should handle spec correctly 1`] = `
"export const addPet = createRequestAction<{
  requestBody: {
    body: IPet;
  };
}>(\\"addPet\\", ({ requestBody }) => ({
  url: \`/api/test/pet\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createItem = createRequestAction<
  {
    requestBody: {
      item?: IItem;
    };
  },
  IMessage
>(\\"createItem\\", ({ requestBody }) => ({
  url: \`/api/test/items\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUser = createRequestAction<{
  requestBody: {
    body: IUser;
  };
}>(\\"createUser\\", ({ requestBody }) => ({
  url: \`/api/test/user\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUsersWithArrayInput = createRequestAction<{
  requestBody: {
    body: IUser[];
  };
}>(\\"createUsersWithArrayInput\\", ({ requestBody }) => ({
  url: \`/api/test/user/createWithArray\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUsersWithListInput = createRequestAction<{
  requestBody: {
    body: IUser[];
  };
}>(\\"createUsersWithListInput\\", ({ requestBody }) => ({
  url: \`/api/test/user/createWithList\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const deleteAttachmentUsingDelete = createRequestAction<{
  id: string;
}>(\\"deleteAttachmentUsingDelete\\", ({ id }) => ({ url: \`/api/test/\${id}\`, method: \\"delete\\" }));

export const deleteOrder = createRequestAction<{
  orderId: number;
}>(\\"deleteOrder\\", ({ orderId }) => ({ url: \`/api/test/store/order/\${orderId}\`, method: \\"delete\\" }));

export const deletePet = createRequestAction<{
  petId: number;
}>(\\"deletePet\\", ({ petId }) => ({ url: \`/api/test/pet/\${petId}\`, method: \\"delete\\" }));

export const deleteUser = createRequestAction<{
  username: string;
}>(\\"deleteUser\\", ({ username }) => ({ url: \`/api/test/user/\${username}\`, method: \\"delete\\" }));

export const downloadUsingGet = createRequestAction<
  {
    id: string;
  },
  IResource
>(\\"downloadUsingGet\\", ({ id }) => ({ url: \`/api/test/\${id}\`, method: \\"get\\" }));

export const findBookByIdUsingGet = createRequestAction<
  {
    id: string;
  },
  IBookDetailVo
>(\\"findBookByIdUsingGet\\", ({ id }) => ({ url: \`/api/test/book/\${id}\`, method: \\"get\\" }));

export const findPetsByStatus = createRequestAction<
  {
    status: keyof typeof FindPetsByStatusStatus[];
  },
  IPet[]
>(\\"findPetsByStatus\\", ({ status }) => ({
  url: \`/api/test/pet/findByStatus\`,
  method: \\"get\\",
  params: {
    status,
  },
}));

/**
 * @deprecated findPetsByTags
 */

export const findPetsByTags = createRequestAction<
  {
    tags: string[];
  },
  IPet[]
>(\\"findPetsByTags\\", ({ tags }) => ({
  url: \`/api/test/pet/findByTags\`,
  method: \\"get\\",
  params: {
    tags,
  },
}));

export const getDocumentByIdUsingGet = createRequestAction<
  {
    documentId: string;
    from?: keyof typeof GetDocumentByIdUsingGetFrom;
  },
  IDocumentVo
>(\\"getDocumentByIdUsingGet\\", ({ documentId, from }) => ({
  url: \`/api/test/documents/\${documentId}/doc\`,
  method: \\"get\\",
  params: {
    from,
  },
}));

export const getHealthCheckUsingHead = createRequestAction<
  {
    flight_no: string;
  },
  IDocumentVo
>(\\"getHealthCheckUsingHead\\", ({ flight_no }) => ({ url: \`/api/test/health-check\`, method: \\"head\\" }));

export const getInventory = createRequestAction<{ [key: string]: any }>(\\"getInventory\\", () => ({
  url: \`/api/test/store/inventory\`,
  method: \\"get\\",
}));

export const getItems = createRequestAction<
  {
    limit?: number;
    page?: number;
    sort?: string[];
  },
  TItems
>(\\"getItems\\", ({ page, limit, sort }) => ({
  url: \`/api/test/items\`,
  method: \\"get\\",
  params: {
    page,
    limit,
    sort,
  },
}));

export const getOrderById = createRequestAction<
  {
    orderId: number;
  },
  IOrder
>(\\"getOrderById\\", ({ orderId }) => ({ url: \`/api/test/store/order/\${orderId}\`, method: \\"get\\" }));

export const getPetById = createRequestAction<
  {
    petId: number;
  },
  IPet
>(\\"getPetById\\", ({ petId }) => ({ url: \`/api/test/pet/\${petId}\`, method: \\"get\\" }));

export const getScheduleDetailsByDateUsingGet = createRequestAction<
  {
    roleId?: string;
    scheduleDate: number;
  },
  IScheduleVo[]
>(\\"getScheduleDetailsByDateUsingGet\\", ({ scheduleDate, roleId }) => ({
  url: \`/api/test/schedules\`,
  method: \\"get\\",
  params: {
    scheduleDate,
    roleId,
  },
}));

export const getTeamListByUsingGet = createRequestAction<{
  limit?: number;
  offset?: number;
}>(\\"getTeamListByUsingGet\\", ({ offset, limit }) => ({
  url: \`/api/test/teams\`,
  method: \\"get\\",
  params: {
    offset,
    limit,
  },
}));

export const getUserByName = createRequestAction<
  {
    username: string;
  },
  IUser
>(\\"getUserByName\\", ({ username }) => ({ url: \`/api/test/user/\${username}\`, method: \\"get\\" }));

export const getUserListByUsingGet = createRequestAction<{
  limit?: number;
  offset?: number;
}>(\\"getUserListByUsingGet\\", ({ offset, limit }) => ({
  url: \`/api/test/users\`,
  method: \\"get\\",
  params: {
    offset,
    limit,
  },
}));

export const loginUser = createRequestAction<
  {
    password: string;
    username: string;
  },
  string
>(\\"loginUser\\", ({ username, password }) => ({
  url: \`/api/test/user/login\`,
  method: \\"get\\",
  params: {
    username,
    password,
  },
}));

export const logoutUser = createRequestAction(\\"logoutUser\\", () => ({ url: \`/api/test/user/logout\`, method: \\"get\\" }));

export const placeOrder = createRequestAction<
  {
    requestBody: {
      body: IOrder;
    };
  },
  IOrder
>(\\"placeOrder\\", ({ requestBody }) => ({
  url: \`/api/test/store/order\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updateBookByIdUsingPut = createRequestAction<{
  id: string;
  requestBody: {
    updateBookRequest: IUpdateBookRequest;
  };
}>(\\"updateBookByIdUsingPut\\", ({ id, requestBody }) => ({
  url: \`/api/test/book/\${id}\`,
  method: \\"put\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updatePet = createRequestAction<{
  requestBody: {
    body: IPet;
  };
}>(\\"updatePet\\", ({ requestBody }) => ({
  url: \`/api/test/pet\`,
  method: \\"put\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updatePetWithForm = createRequestAction<{
  petId: number;
  requestBody: {
    name?: string;
    status?: string;
  };
}>(\\"updatePetWithForm\\", ({ petId, requestBody }) => ({
  url: \`/api/test/pet/\${petId}\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export const updateUser = createRequestAction<{
  requestBody: {
    body: IUser;
  };
  username: string;
}>(\\"updateUser\\", ({ username, requestBody }) => ({
  url: \`/api/test/user/\${username}\`,
  method: \\"put\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const uploadAttachmentUsingPost = createRequestAction<
  {
    requestBody: {
      attachment: File;
    };
  },
  IAttachmentBo
>(\\"uploadAttachmentUsingPost\\", ({ requestBody }) => ({
  url: \`/api/test\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export const uploadFile = createRequestAction<
  {
    petId: number;
    requestBody: {
      additionalMetadata?: string;
      file?: File;
    };
  },
  IApiResponse
>(\\"uploadFile\\", ({ petId, requestBody }) => ({
  url: \`/api/test/pet/\${petId}/uploadImage\`,
  method: \\"post\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export interface IApiResponse {
  code?: number;
  message?: string;
  type?: string;
}

export interface IAttachmentBo {
  authorName?: string;
  createdDate?: number;
  fileName?: string;
  id?: string;
  mimeType?: string;
  path?: string;
}

export interface IBookDetailVo {
  attachment?: IScheduleVo;
  authorName?: string;
  createdDate?: number;
  fileName?: string;
  id?: string;
  mimeType?: string;
  path?: string;
}

export interface IBookVo {
  address?: string;
  price?: string;
}

export interface IBookingResponse {
  data: IDocumentVo;
  errors?: IErrorInfo[];
}

export interface ICategory {
  id?: number;
  name?: string;
}

export interface IDocumentVo {
  attachment?: IBookDetailVo;
  authorName?: string;
  createdDate?: number;
  id?: string;
  note?: string;
  title?: string;
}

export interface IError {
  code?: number;
  field?: string;
  message?: string;
}

export interface IErrorInfo {
  errorMessage?: string;
}

export type TErrors = IError[];

export interface IFile {
  absolute?: boolean;
  absoluteFile?: IFile;
  absolutePath?: string;
  canonicalFile?: IFile;
  canonicalPath?: string;
  directory?: boolean;
  executable?: boolean;
  file?: boolean;
  freeSpace?: number;
  hidden?: boolean;
  lastModified?: number;
  name?: string;
  parent?: string;
  parentFile?: IFile;
  path?: string;
  readable?: boolean;
  totalSpace?: number;
  usableSpace?: number;
  writable?: boolean;
}

export enum FindPetsByStatusStatus {
  \\"available\\" = \\"available\\",
  \\"pending\\" = \\"pending\\",
  \\"sold\\" = \\"sold\\",
}

export enum GetDocumentByIdUsingGetFrom {
  \\"AAA\\" = \\"AAA\\",
  \\"BBB\\" = \\"BBB\\",
}

export interface IInputStream {
  [key: string]: any;
}

export interface IItem {
  color?: string;
  id?: string;
  name?: string;
  price?: string;
}

export type TItems = IItem[];

export interface IMessage {
  message?: string;
}

export interface IOrder {
  complete?: boolean;
  id?: number;
  petId?: number;
  quantity?: number;
  shipDate?: string;
  status?: keyof typeof OrderStatus;
}

export enum OrderStatus {
  \\"placed\\" = \\"placed\\",
  \\"approved\\" = \\"approved\\",
  \\"delivered\\" = \\"delivered\\",
}

export interface IPet {
  category?: ICategory;
  id?: number;
  name: string;
  photoUrls: string[];
  status?: keyof typeof PetStatus;
  tags?: ITag[];
}

export enum PetStatus {
  \\"available\\" = \\"available\\",
  \\"pending\\" = \\"pending\\",
  \\"sold\\" = \\"sold\\",
}

export interface IResource {
  description?: string;
  file?: IFile;
  filename?: string;
  inputStream?: IInputStream;
  open?: boolean;
  readable?: boolean;
  uri?: IUri;
  url?: IUrl;
}

export interface IScheduleVo {
  schedules?: IBookVo[][];
  shiftId?: string;
  team?: string;
}

export interface ITag {
  id?: number;
  name?: string;
}

export interface IUpdateBookRequest {
  birthCountry?: string;
  citizenship?: string;
  dateOfBirth?: number;
  employmentStatus?: string;
  ethnicity?: string;
  gender?: string;
  idNumber?: string;
  idType?: string;
  roleId?: string;
  spokenLanguage?: string[];
}

export interface IUri {
  absolute?: boolean;
  authority?: string;
  fragment?: string;
  host?: string;
  opaque?: boolean;
  path?: string;
  port?: number;
  query?: string;
  rawAuthority?: string;
  rawFragment?: string;
  rawPath?: string;
  rawQuery?: string;
  rawSchemeSpecificPart?: string;
  rawUserInfo?: string;
  scheme?: string;
  schemeSpecificPart?: string;
  userInfo?: string;
}

export interface IUrl {
  authority?: string;
  content?: { [key: string]: any };
  defaultPort?: number;
  deserializedFields?: IUrlStreamHandler;
  file?: string;
  host?: string;
  path?: string;
  port?: number;
  protocol?: string;
  query?: string;
  ref?: string;
  serializedHashCode?: number;
  userInfo?: string;
}

export interface IUrlStreamHandler {
  [key: string]: any;
}

export interface IUser {
  email?: string;
  firstName?: string;
  id?: number;
  lastName?: string;
  password?: string;
  phone?: string;
  userStatus?: number;
  username?: string;
}

export interface IValidationError {
  errors?: TErrors;
  message?: string;
}
"
`;
